### ISSUE TITLE
Require a user gesture for a cross-origin frame to navigate the top-level page

### ISSUE THREAD

**Issue by ojanvafai** [WICG/interventions#16]
We are seeing a lot of ads that navigate the top-level page outside of a user gesture. We think this is user hostile and that we don't lose any important use cases by requiring a user gesture. Chrome has added formal measurement for this. Once we get the numbers back, we plan to make this change.

There is also the allow-top-navigation flag for iframe sandbox attribute. Once we make this change the plan is that sandbox disallows top navigation and allow-top-navigation allows the behavior described above (i.e. even with allow-top-navigation, you'd still need a user gesture to navigate the top-level page).

**Comment by MattMenke** [comment]
Any consideration of doing this for modal dialogs as well?  When I saw the email about not opening "popups" except in response to a user gesture, was excited about this because I thought it included blocking that particular behavior as well.  A lot of abusive pages use OnBeforeUnload dialogs, which are just annoying.

**Comment by mikewest** [comment]
blink-dev thread: https://groups.google.com/a/chromium.org/d/msg/blink-dev/Xi8-y4ySjA4/dc4BFfqyBQAJ

/cc @annevk @rlbmoz @johnwilander @travisleithead

**Comment by hillbrad** [comment]
:(  This breaks a thing I'm working on.  I'm trying to expose a cross-origin API in an iframe via postMessage.  The parent frame uses postMessage to request a service from the iframe.  If the request can be fulfilled without user confirmation, the iframe navigates the parent to a "success" page specified by the caller. Otherwise it makes a POST that navigates the parent frame to a confirmation page, which takes the user back to the specified "success" page after the user gesture.

I like the iframe top navigation pattern for a couple of reasons.

Compare it to the traditional workflow for doing this (as with OAuth) which requires POSTing directly to the cross-origin resource or opening a popup:
* If user consent isn't required the user is taken directly to the "success" page without incurring a synchronous additional round-trip and redirect in the top frame, giving better perceived performance, responsiveness and less drop-off.
* Popups suck and are frequently blocked. 

The parent could implement another listener for a postMessage reply and do the navigation itself, but that makes using this kind of API more complicated, and it is nice to be able to do the POST from the iframe and include, e.g. a CSRF token and trustworthy indication of the parent origin that triggered the request.

A click to the iframe doesn't make sense; it takes away control over the UX that the parent wants.

Since a page that doesn't want this behavior from an iframe already has the ability to block top-navigation with the sandbox attribute, I'm not sure why we need to break this behavior by default?

Could we count receiving a postMessage from the parent as enabling navigation? (in addition to a user gesture)

**Comment by manki** [comment]
This is breaking sign-in flow for CapitalOne (https://servicing.capitalone.com/c1/Login.aspx). I see the following message in console:

    Unsafe JavaScript attempt to initiate navigation for frame with URL 'https://servicing.capitalone.com/c1/Login.aspx' from frame with URL 'https://login2.capitalone.com/loginweb/login/postLogin.do'. The frame attempting navigation is targeting its top-level window, but is neither same-origin with its target nor is it processing a user gesture. See https://www.chromestatus.com/features/5851021045661696.

I presume Capital One will have to fix their site, and Chrome is working as intended?

**Comment by reginaldos** [comment]
We received user reports that this is breaking users of the Facebook SDK. https://developers.facebook.com/docs/games/gamesonfacebook/login explicitly encourages use of top level navigation.

**Comment by ojanvafai** [comment]
@reginaldos thanks for the report. Can you point us to an example page? We considering changing behavior here and being able to verify the new idea works would be helpful. It's hard for me to follow in that SDK documentation where top-level navigation is used.

**Comment by jontonsoup** [comment]
@ojanvafai this also seems to break typeform redirection

**Comment by loganrosen** [comment]
This broke eBay checkout flow for me. It hung after adding a credit card in PayPal, and an error in the console pointed me to this new Chrome feature.

In the console:

> main.js:21966 Unsafe JavaScript attempt to initiate navigation for frame with URL 'https://mbuy.ebay.com/xo?action=view&sessionid=(redacted)&redirect=mobile' from frame with URL 'https://www.paypal.com/webapps/helios?action=addCard&payer_id=(redacted)…hod%3DPAYPAL%26redirect%3Dmobile#/checkout/pageAddCard/addCardFlow/addCard'. The frame attempting navigation is targeting its top-level window, but is neither same-origin with its target nor is it processing a user gesture. See https://www.chromestatus.com/features/5851021045661696.
> 
> findWindowAndRedirect @ main.js:21966

**Comment by natechapin** [comment]
I just reenabled this experiment in Chrome 56+ builds (should be on the chrome canary in a day or two) with a few changes to try to make it less breaking:
* Instead of requiring a current user gesture, a cross-origin iframe can navigation top if it has ever received a user gesture.
* When setting the bit that indicates whether a given browsing context has received a user gesture, it is also set on all of that browsing context's ancestors.

**Comment by hillbrad** [comment]
I still think that the parent frame sending a postMessage should also
"activate" the iframe, or there needs to be a way to explicitly allow a
child frame to top-nav so without a user gesture.  This is an important
cross-origin API surface.  Iframes are not just ads.

**Comment by ojanvafai** [comment]
@hillbrad is it possible for your case to postMessage back to the parent page to do the navigation?

Having postMessage auto-grant this ability would take all the value out of this proposal since postMessage is so widely used.

Maybe we should consider adding a permission that we can delegate down via feature policy (e.g. via an enable attribute on the iframe) to allow the top-level page to say the iframe can navigate it without a user gesture. @hillbrad would that work for your use case?

**Comment by hillbrad** [comment]
@ojanvafai an explicit opt-in or callback could work, but there are 7 independent bug reports here, which seems to indicate reasonably wide legitimate use.

What does the data look like on the prevalence of legitimate vs. user-hostile occurrences?

**Comment by ojanvafai** [comment]
@hillbrad these bug reports were on the less permissive version that required a user gesture for the navigation instead of any user gesture during the life of the frame. It's hard to test all of these because most of them don't link to pages, but I confirmed that the capitalone case does get a user gesture today.

It's very hard to get data on prevalence of legitimate vs. user-hostile at scale. If we could distinguish the two reliably, we'd do a different intervention. :)

If we get reports of real content that breaks with this more permissive solution, we'll need to reconsider.

**Comment by hillbrad** [comment]
It seems that any use of hidden frames as an API surface that includes
automatic navigation will still break under the new approach because they
have no way to collect a user gesture?  So I think there should be a way to
designate a frame at creation as capable of top navigating, or to decorate
a postMessage so it delegates a capability equivalent to a user gesture,
before intervening and requiring a larger redesign of these existing uses.
Setting up an event listener, parsing and navigating in response to a reply
message is not terribly difficult, but it is much more complicated than a
simple fire-and-forget postMessage.

**Comment by ojanvafai** [comment]
I think a way to delegate top navigation to subframes is reasonable. It would probably build on top of feature policy: https://wicg.github.io/feature-policy/. When we ship feature policy, I don't see any downside to adding the ability to delegate down the ability to do top-navigation without a gesture. But, I'm hesitant to block shipping this until then due to:
1. The large user benefit this has.
2. (so far) no complaints of real-world web sites breaking.
3. The restriction can be worked around via a postMessage to the parent.

I do think we might want to explore having a postMessage that happens during a user-gesture grant the messaged-frame the ability to top-navigate since we pass the user gesture bit with the postMessage already (e.g. the inner frame's message handler can open popups, etc). That'd probably be good for keeping the platform simple and consistent. But, I think that doesn't actually address your need @hillbrad, right?

**Comment by hillbrad** [comment]
If you can track a user gesture in a parent frame leading to an postMessage
to the child frame and carry the user intent flags through, that should
work for my specific use cases.  But it would definitely be nice to have
something deterministic to set so every postMessage API documentation
doesn't need an extra page of explanation about this behavior.

**Comment by limpygnome** [comment]
I'm seeing this in Chrome 56, but not for post message. Does anyone know if this will definitely ship on January 31st? I am not familiar with the Chromium release cycle, and whether dates are strictly followed.

**Comment by ojanvafai** [comment]
@limpygnome the dates aren't strictly followed, but it should ship on January 31st give or take a week. Is this breaking a page for you?

**Comment by jitdasrjs** [comment]
In our facebook application we can not redirect pages using top.location.href any more in chrome v56 beta.
It worked fine in previous version.Now javascript error is coming-The frame attempting navigation is targeting its top-level window, but is neither same-origin with its target nor is it processing a user gesture.

**Comment by tonygaillard** [comment]
https://apps.facebook.com/candycrush is broken on first install
https://apps.facebook.com/candycrushsoda is broken on first install
https://apps.facebook.com/livepool/ is broken on first install

**Comment by uofifox** [comment]
I have several applications that I work with that will be broken with this new feature.  We have requirements for us to send users out to external websites and redirections.  Why would you block an internal application from making decisions about where it should go?

In software if you don't provide a way for either a sub-element to communicate with the parent to help make informed decisions, or allow the child element to make a decision, you're in essence making the child restricted to basically a small task.  There are many situations where the child is better suited to make decisions related to the situation it is being used and is basic to decomposing systems.  Almost like inheritance, and saying the inherited element is allowed to only do a couple of things, but you can't add functionality.

Also note that many systems have clearly been built using iframes as modular breakdowns.  Hopefully this feature will be augmented or an api will be added allowing the child frame to communicate with the parent otherwise I forsee chrome being an issue for many users moving forward.

**Comment by domenic** [comment]
> an api will be added allowing the child frame to communicate with the parent

This sounds like the already-existing postMessage API

**Comment by uofifox** [comment]
Valid point, looks like the postMesage API could be used to accomplish the issue.  Of course means the parent frame would need to add a listener to the events.  Thank you for the reference.

**Comment by ojanvafai** [comment]
@tonygaillard this looks like a separate bug. Those apps are broken with this change disabled as well. Filed crbug.com/674229 to track that.

@uofifox sounds like your issue is resolved?

**Comment by uofifox** [comment]
Well I would prefer the feature not get implemented as opposed to saying my issue is resolved. In my case it's great there is a workaround but as I don't have control over the main frame it means we have to ask for feature enhancements which take time to implement

**Comment by joeedh** [comment]
What is a "user gesture"?

**Comment by uofifox** [comment]
User initiated action

**Comment by tonygaillard** [comment]
@uofifox Nope, this is this bug/feature, because the redirection in JavaScript is not allowed. Go to either one of the application and open the developer console to notice the issue.

**Comment by jerodvenemafm** [comment]
This is impacting a site of mine as well, which has an iframe-based login page. We do a login in the iframe via SAML (don't ask), and then on completion, a postMessage to signal the parent to navigate, which fails under the new rules. This is on the same root domain (different subdomains). I'd be OK with a frame-level setting to enable the behavior, or enabling via postMessage.

**Comment by elexisvenator** [comment]
This issue also impacts the use of the SecureFrame payment api from SecurePay. After payment is made, SecurePay is meant to redirect to a url on your site to notify your application of successful payment. This never happens and results in payments being made without the application knowing about it (reconciliation nightmare).

I have pointed this out to the SecurePay devs, who say they will monitor this feature.

SecurePay SecureFrame integration guide for reference: https://www.securepay.com.au/_uploads/files/SecureFrame_Integration_Guide.pdf

**Comment by chrisbro-MSFT** [comment]
This also breaks login flows for Microsoft Office Online.  Example flow:

1. A Dropbox for Business user clicks on a .docx file in the Dropbox web UI.
2. Dropbox iframes the Word Online app which runs on a different domain.
3. Word Online checks for the user's login cookie and, not finding it, needs to redirect them to the Office 365 login page.  Login pages can't be framed, of course, so it needs to navigate window.top to the login experience.  This breaks in Chrome 56.

Here's some documentation on how that flow works: https://wopi.readthedocs.io/en/latest/scenarios/business.html#validating-edit-capabilities

We see two workarounds right now: 
1. We could deliver a little blob of HTML that says "hey buddy you need to log in, click here" and the user would click that to go to the page that asks them to log in.  Degrades the user experience by forcing the user to click instead of redirecting them automatically so we'd only do this for Chrome users, but it would technically work and will be our plan if you don't provide another solution.
2. If you implement a way for the host page to opt-in to this behavior, we could probably get all of our partners to implement the workaround by end of month.  It only breaks their paying business users so they'll be motivated to work quickly...

If you'd like to discuss this in email, my email address is my username at microsoft.com.

**Comment by lubin2010** [comment]
Working on a relevant effort #42: "Require a user gesture for sandboxed iframe w/ 'allow-top-navigation' to navigate the top-level page" which has a similar purpose as this one but covers different use cases and might break less pages and thus less risky. Any concern with it?

**Comment by chrisbro-MSFT** [comment]
@lubin2010 - thanks, no immediate concerns as I don't think any hosts load our apps via sandboxed iframes, and if we do find issues the workaround is clear.

**Comment by limpygnome** [comment]
@ojanvafai yes, this breaks our payment pages for an iframe integration, whereby it needs to redirect the parent page. Our product processes a large volume of transactions for tier one/two merchants, so we're watching this closely and releasing a fix before the scheduled Chromium release date.

For now we've fixed it using post message, tested against Chrome 56 dev. This does not require any user interaction. Would this work-around be broken in the future?

**Comment by Hammadk** [comment]
[Shopify's](https://shopify.com) embedded app SDK requires top level navigation with [window.top.location.href](https://help.shopify.com/api/sdks/embedded-app-sdk/getting-started#oauth) during OAuth. This is similar to the Facebook's SDK and many other platforms that assume this functionality -- this change will break logins on a lot of platforms.

This is a reasonable change (to prevent malicious ads) but it should be delayed and at least have a deprecation warning that Chrome developers (non-Canary) would see. That'll give developers more time to come up with solutions.

**Comment by ojanvafai** [comment]
@jerodvenemafm I don't quite understand your example. postMessage to the parent and then the parent navigating the top-level page should be unaffected by this change. Do you have a page you can point us to so we can understand what's going on?

@elexisvenator can you point to an example that breaks? Reading through the pdf it seems like the user would be require to interact with the frame?

@limpygnome using postMessage should be totally safe. The point of this change is to make it so that the top-level page is in control of navigations. If the top-level page does the navigation in response to a postMessage, then it's still in charge. :)

@Hammadk, do you have a page you could link us to that we can test?

For now, I think we'll probably change Chrome to show a console warning instead of blocking while we evaluate our options here. Thanks all for the feedback! In general, pointing us to actual sites that break is most helpful because then we can try our best to design solutions that don't break them whenever possible.

**Comment by jerodvenemafm** [comment]
URL is private. But it's part of the login process. 

Iframe login->POST->successful login->postMessage to parent->parent redirects

The iframe resides on a different sub-domain than the main app.

**Comment by Hammadk** [comment]
Thanks Ojan. Shopify's apps are private and available to merchants with online stores. If you have trial store, you can install embedded apps to test this change:
https://YOUR-SHOP.myshopify.com/admin/apps/shopify-widgets

If you don't have a Shopify store, our setup is pretty similar to Facebook apps. You can test this change by installing a Facebook app:
https://apps.facebook.com/livepool/

**Comment by elexisvenator** [comment]
Ojan, 
The last step of the SecurePay payment process is to confirm payment via clicking a button. This redirects the iframe to a "payment is processing" page which after a while redirects the parent to the payment complete page. It is this last redirect that appears to fail.

As for examples, I am afraid our application doe not have any publicly accessible sites with the test payment interface configured. Securepay also does not appear to have a public demo. An example of an actual live site would be shopmate.auspost.com.au (auspost is their parent company).

**Comment by adob** [comment]
:-1: :-1: This change appears to be causing a lot of breakage in my end-to-end Selenium/WebDriver because a lot of the interaction is not treated as user-initiated.

**Comment by limpygnome** [comment]
@ojanvafai unfortunately we've had a few issues with using an alternative in such a short time period.

Can you definitely confirm whether Chrome 56 will be changed to feature a console message instead? I think we're still seeing the blocked behaviour on dev.

**Comment by lubin2010** [comment]
@limpygnome AFAIK, yes. The code was submitted and will be merged in Chrome 56:
https://bugs.chromium.org/p/chromium/issues/detail?id=678328

**Comment by ojanvafai** [comment]
We're working on reverting it in chrome 56.

For those of you who have commented on breakage, since we can't test sites we don't have access too, it would be really helpful if you could tell us if proposed next attempt will still break your pages. Also, we should have it behind a flag in a week or two for you to test.

The Chrome 56 behavior requires the document to have ever had a user gesture. Which means that the bit is lost when the iframe is navigated. For Chrome 57, we're going to try the same thing, but will make it only require that that iframe had a user gesture ad some point. Namely, it will survive navigations in that iframe.

This will still break cases where an iframe is completely hidden from the user and the user never clicks or types in it.

Incidentally, I ran into https://news.ycombinator.com/item?id=13356074 today as an example of the sort of problem this change will mitigate.

**Comment by riking** [comment]
> since we can't test sites we don't have access to

Here's one that should be easy to test: https://www.twitch.tv/products/bits/B018WMZN5E

UI path to that URL:

 - Create an account on www.twitch.tv
 - Go to any partnered channel with Bits enabled
 - Attempt to buy $5.00 worth of Bits ($7.00 button)
 - Popup gets error

>Unsafe JavaScript attempt to initiate navigation for frame with URL 'https://www.twitch.tv/products/bits/B018YYYYYY?tv=eyJraWQiOiJjb20uYW1hem9uL…BlvswUhD7tKzLMVWz91pSgc_mvfn0id23MY2w' from frame with URL 'https://twitch.amazon.com/checkout/summary?embed=true&asin=B018YYYYYY&tuid=eyJraWQiOiJjb20uYW1hem9uL…BlvswUhD7tKzLMVWz91pSgc_mvfn0id23MY2w'.

**Comment by mikewoo200** [comment]
Here is the eBay Checkout flow that was broken in Version 56.0.2924.51 beta (64-bit) - before this change was reverted recently.

1) Go to ebay (www.ebay.com) and sign in
2) Buy any item to get to checkout page. I used item id 121409414417

http://www.ebay.com/itm/eBay-Mobile-QA-10-0-Test-Item-12-/121409414417

3) In Checkout page, click on PayPal or PayPal Credit to open the PayPal sign-in layer
4) Without logging into PayPal, close the PayPal layer that came up
5) The Unsafe JS error shows in console log. User is stuck with the page mask, unable to use the page
6) User refreshes eBay checkout page but user is NOT signed into PayPal
7) User who wants to pay with PayPal is now stuck in infinite loop

Thanks.

**Comment by linkkingjay** [comment]
This also breaks WeChat OAuth login flows. Example flow:

1. A web application use an iframe to insert a QR code in its login page.
1. User use WeChat app in his/her phone to scan the QR code.
1. After confirming login in WeChat app, the iframe will redirect the parent to home page.

https://www.weiyun.com/

POC:

```javascript
success() {
  window.top.location = redirect_url
}
```

![image](https://cloud.githubusercontent.com/assets/2505538/21976583/8489f534-dc0d-11e6-8e63-1929c8af8f31.png)

**Comment by limpygnome** [comment]
Do you have any detail around how this will be more "permissive" in Chrome 57?

**Comment by mikewoo200** [comment]
Will the "more permissive" version come in version 57 still? If so, when will we get a beta version to test?

**Comment by natechapin** [comment]
@limpygnome In Chrome 57, if an iframe has ever received a user gesture, then the iframe can navigate top. The reverted experiment in Chrome 56 reset this state when the iframe navigated to a different document.

@mikewoo200 Yes, the more permissive version should be available in Chrome 57 builds (currently on canary and dev channels, not sure when the 57 beta will ship).

**Comment by lubin2010** [comment]
The Chrome 57 beta has been released to Desktop and Android (see [the blog post](https://chromereleases.googleblog.com/)), and it can be downloaded [here](https://www.chromium.org/getting-involved/dev-channel). Please feel free to test it and report any issue you find. Thanks.

**Comment by mikewoo200** [comment]
1)What's the scheduled release date of Chrome 57 with this "more permissive" feature?
2) Is this particular feature considered complete for Chrome 57?

One version of eBay checkout is still broken with this "more permissive" version (but broken differently from 56 Beta), with this error:

```The frame attempting navigation is targeting its top-level window, but is neither same-origin with its target nor has it received a user gesture. See https://www.chromestatus.com/features/5851021045661696.```

We will work on a fix using this version of Chrome 57 Beta.

**Comment by lubin2010** [comment]
@mikewoo200 Could you share how to reproduce that broken checkout process on eBay? I couldn't reproduce it with the steps you mentioned [above](https://github.com/WICG/interventions/issues/16#issuecomment-272232372).

**Comment by mikewoo200** [comment]
@lubin2010, Here are new steps for Chrome 57 Beta.

1. Go to ebay (www.ebay.co.uk) and sign in. (Notice it's UK site)
2. Buy any item to get to checkout page. I used item id 121409414417
http://www.ebay.co.uk/itm/eBay-Mobile-QA-10-0-Test-Item-12-/121409414417
3. Once the UK checkout page is loaded, click on PayPal or PayPal Credit to open the PayPal sign-in layer (pop up window)
4. Without logging into PayPal, close the PayPal layer that came up
5. The Unsafe JS error shows in console log. User is stuck with the page mask, unable to use the page
6. User refreshes eBay checkout page but user is NOT signed into PayPal
7. User who wants to pay with PayPal is now stuck in infinite loop

Thanks.

**Comment by MrStonedOne** [comment]
It seems like this should be opt in on the site, not opt out.

I'm failing to understand why breaking the flow of every oauth system out there is necessary.

**Comment by MrStonedOne** [comment]
Or actually, why are we not giving a topbar or a redirection blocked ui (like the popup blocked ui) to give the user the ability to say no, go ahead to that site.

**Comment by ziyunfei** [comment]
An opt-out should be provided, but personally, I don't want a popup blocking UI 😜.

**Comment by MrStonedOne** [comment]
except heres the thing, I'm encountering this for my oauth workflow, only i don't have the ability to work around it. So allowing the user to approve the redirection is the only way my shit will work.

**Comment by chrcoluk** [comment]
found this page after seeing this flag was toggled to an enabled state in chrome://flags when I didnt make the change myself, is chrome now automatically enabling some experimental flags now?

**Comment by aels** [comment]
@chrcoluk, yes, all "Default" state flags can be toggled to "On" by chrome/google guys at any time.

**Comment by mexhanic** [comment]
@ojanvafai 
@lubin2010 
You want to hit all rabbits with one bullet, but what about granular solution?
We have one thing, that brokes - oAuth, and one thing that must be stoped - ads framebusting.

So, lets assume we whitelist of (for example) "top Alexa 1k" domains, which allowed to bust-out their cross-domain top's. But let them still throw console warning from frame page, to notice developers.

If our frame is not in whitelist, let assume following algo when frame page try to bust-out (and cross-domain, and no gesture given):
- issue a warning to console from frame page (no notice frame-page developers)
- show access-request dialog on top page (like camera access request)
- issue a warning to console from *top* page (with url of frame-initiator), to allow in-page reporters to log, that someone wants to redirect us.
- if "granted", then pause redirection for 2 seconds and issue a warning from top page, to allow of logging this redirection. Then redirect.
- if "declined", then throw error to frame page, and issue warning to top page, to allow top-page developers to log incident.

Mechanism of "warnings" is on your own. But I think it will be good to include not only frame url, but urls of all frames in chain, in case of nested frames.
That raises second problem - possible leakage of any secure tokens inside intermediate urls to the top-page. But that needs additional research.

**Comment by mexhanic** [comment]
And second idea.
Mostly all of ad-frames have specific size (like 300x250), when oAuth-frames haven't.
So, if any of frames in chain from redirect-initiator to top page have ad-like size, then block.

**Comment by lubin2010** [comment]
We've reverted the change in Chrome 57, and are rethinking how to go forward with this, and will keep you updated once there is more to share. Thanks all for the feedback!

**Comment by oniric85** [comment]
Today I saw the warning the first time on my Chrome Console. We have a login system based on OAuth that automatically redirects to a success page by changing top navigation from inside the iframe. This breaks now that I activated the flag in Chrome flags as it does not require any user gesture on certain condition to increase responsiveness and performances.

Is this moving forward? Is there any way to whitelist the behavior from the top page?

**Comment by lubin2010** [comment]
+ @ojanvafai to comment on the status of this.

**Comment by oniric85** [comment]
Any news? We could really make use of some rough timings as to adapt to this change as we would need to plan work on several systems that currently rely on this feature, probably switching to postMessage.

**Comment by tomloake** [comment]
I noticed some of the work to unblock this has happened recently (main frame navigation consumes a user gesture) has started to move forwards - do we have any updates on this status of this bug now?

**Comment by MrStonedOne** [comment]
This issue is, (and will always be) detection.

A script (be it a good faith one like sso/oauth, or a bad faith one like ads) that fails to meet the criteria will have no way of knowing before hand, and will have its entire frame errored out, leaving no way to recover from the error or fallback to a gesture request.

**Comment by lubin2010** [comment]
It was [announced](https://blog.chromium.org/2017/11/expanding-user-protections-on-web.html) that in Chrome 64 (To be the stable version in around late Jan, 2018), all redirects originating from third-party iframes will show an infobar instead of redirecting, unless the user had been interacting with that frame. 
 
+@rschoen to answer questions if there are any.

**Comment by MrStonedOne** [comment]
Is there a way to determine, at the script level, if a topframe redirect would lead to an infobar.

Also how does the blocking chain on that work? would the location set block while the infobar is up, what happens if it is denied, does the frame error out (like it did before) or does the code continue on as normal?

I have to use this in an oauth flow that involves in game prompts because of the way the game engine presents these things to users and the fact the devs refuse to create a native oauth way to authenticate users. I can present them with a button, but the system works best if it can auth them in the background during the loading and redirect them once both complete.

**Comment by chrisbro-MSFT** [comment]
@lubin2010 @rschoen, the announcement does not describe how the allow-top-navigation and allow-top-navigation-by-user-activation flags are impacted with this current change, and the announcement does not explain how a site can opt-in to actionless navigation in a particular iframe.  Could you please clarify?

**Comment by Mehul24** [comment]
>Restrict this ability to content that is processing a user gesture, unless it is same-origin to the parent.
 
Can an iframe(on a different domain) not change the top location even though its changing the domain to the same domain as the parent(top) domain? The wording is not very clear here 'same-origin to the parent', does that mean the iframe needs to be the same origin as the parent or it can only redirect to the same origin as the parent? 

Example - 
Window = DomainA.X
Iframe = DomainB.y
Iframe changes window = DomainA.Z without any user gesture.

**Comment by mayoque** [comment]
I am a little bit confused about current behavior and I am not sure it was really intended to work this way.
At this time (v69) cross-origin redirection within top window still works even without user gesture once iframe is sandboxed with "allow-top-navigation allow-same-origin".
However it does not work when iframe is defined without sandbox attribute.
Is there some intention to make it work ONLY in sandboxed iframe with explicit permissions?
Is it OK it doesn't work in iframe without restrictions - without sandbox attribute?

**Comment by antbewes** [comment]
Since we took the auto update to Chrome from 74.0.3729.169 to 75.0.3770.100 our iframe based payment process no longer works. The iframe (on a different domain) has JS which posts a form to target _top and this used to work. Now we get JS error "Unsafe JavaScript attempt to initiate navigation for frame with URL...", I've managed to fix with explicit sandbox="//every allow option//" attr on the iframe, but can't explain why we're only getting this in v75?

**Comment by natechapin** [comment]
antbewes@. I'm a little surprised that this used to work...

My best guess for why this changed: in Chrome 74->75, I did a bunch of cleanup around targeting windows by name. Among other things, I fixed some bugs where instead of blocking a navigation that targeted a disallowed window, Chrome would instead navigate the wrong window. Is there any possibility that, in Chrome 74, that navigation was executed in the iframe that posted the form, rather than in _top? Skimming through the logic as it stood in 74.0.3729.169, it's quite possible that this case was ending up in the wrong window, where we had intended for it to be blocked.

Regarding the workaround: if any sandbox options besides "allow-top-navigation" are necessary to get this to work, I'd be interested in knowing which ones to make sure there aren't any lingering bugs here.

**Comment by antbewes** [comment]
@natechapin Thanks for responding.
I'm not able to only set "allow-top-navigation" because our iframe needs to "allow-scripts" and "allow-forms" at least to function before it posts its form to target="_top". Because this worked without any sandbox attr pre v75, I took that to mean that if I now have to add a sandbox option then all should be added to preserve the permissioning. Strange though that Chrome 75 should treat iframe with no sandbox attr differently to iframe with all sandbox attr options.

**Comment by natechapin** [comment]
@antbewes yeah, there's some historical quirks in Chrome's navigation logic around sandbox with allow-top-navigation specifically. It's on my backlog... :/

**Comment by ShashankM3-hexaware** [comment]
The frame attempting navigation is targeting its top-level window, but is neither same-origin with its target nor has it received a user gesture. 
I'am facing above mentioned error when I try to get a fresh token from a Microsoft account using MSAL angular package. 
![image](https://user-images.githubusercontent.com/38096199/73534737-d4b1a480-4447-11ea-96cb-c3271f223d2f.png)

**Comment by ShashankM3-hexaware** [comment]
The frame attempting navigation is targeting its top-level window, but is neither same-origin with its target nor has it received a user gesture.
I'am facing above mentioned error when I try to get a fresh token from a Microsoft account using MSAL angular package.
![image](https://user-images.githubusercontent.com/38096199/73534983-8d77e380-4448-11ea-808d-7afc980a07b9.png)

**Comment by MiguelQueiroz** [comment]
Our Stripe payment iframe no longer works, the redirect is no longer happening, blank page :(

**Comment by jessejacob** [comment]
I have a stripe integration with wordpress that is failing because of this. I'm using postMessage to trigger form submission in a child shopping cart iframe and I get blocked by "Unsafe JavaScript attempt to initiate navigation for frame with origin 'https://a.com' from frame with URL 'https://b.com'. The frame attempting navigation is targeting its top-level window, but is neither same-origin with its target nor has it received a user gesture. See https://www.chromestatus.com/features/5851021045661696."
I was expecting postMessage to work after reading some of the discussion above from years back. I also tried sandboxing the child iframe with allow-top-navigation (and then everything) and that didn't work either (same Chrome block). I'm also not seeing the discussed support for either a click that results in a postMessage from parent-to-iframe as a valid user gesture for top navigation from iframe, or an "allow" attribute for iframes to implement a specific permissions policy for top-level navigation. sandbox="allow-top-navigation" basically seems broken at the moment and there is no way around it. Super frustrating.

**Comment by annevk** [comment]
Are there plans to standardize this in the HTML Standard?

**Comment by Fjan** [comment]
Since Chrome version 102 it appears a redirect inside an iframe resets the awareness of previous user gestures in that frame. This means another user gesture is needed before a script can escape from the frame. This broke our app which performed a simple `top.location.href = location.href` on a new page after the user submitted a form.

Mainly putting this out here to help other people running into it to figure out what's going on, but it would be nice if we could use something like `allow="top-navigation: *"` instead of having to put more `onclick` handlers in our code.

**Comment by adamlavery** [comment]
Chrome/Edge v102 has broken PayPal hosted payments. This is a major issue. Hosted payments loads a Pay with PayPal button and card payment form:

![image](https://user-images.githubusercontent.com/13659750/173779621-528b3e78-dfc4-47df-90ca-2c41023203ed.png)

Since Chrome/Edge v102 clicking Pay with PayPal is blocked. 

![image](https://user-images.githubusercontent.com/13659750/173780222-d8c8bd44-8c14-4b44-8d0c-496d0521c222.png)

If clicking the button to go to PayPal is not a user gesture then what is?

The only way I have found to fix this is to sandbox the iframe, but then that breaks card payments. Even with all sandbox restrictions disabled, 3DS processing fails. So we cannot use sandboxed iframes.

Clearly latest Chromium user gesture detection is flawed - how can clicking a button in an iframe not be a user gesture?

Please reverse this change soonest until you can come up with something that works properly.

**Comment by domenic** [comment]
With regard to the original proposal here, the route we went down was to only introduce this restriction to sandbox iframes, and it can be turned off via a variety of flags, such as `allow-top-navigation` or `allow-top-navigation-by-user-activation`. Those cases are specified in https://html.spec.whatwg.org/#allowed-to-navigate and this is [fully cross-browser behavior by now](https://caniuse.com/mdn-html_elements_iframe_sandbox-allow-top-navigation-by-user-activation).

In all cases, this is a choice by the embedding website; if the embedding website wants to allow navigation of the top-level page, it can easily do so, either by not using `sandbox=""`, or by using those flags to reenable the functionality after using sandbox. So, if you are encountering this issue, a solution is to change your embedding website's sandboxing strategy.

However, it appears this issue is attracting a lot of comments unrelated to the original proposal. Basically, any time some part of the web platform causes a navigation failure, people are posting here. Even if it has nothing to do with the sandboxing restriction that was implemented. For example, it seems per https://chromestatus.com/feature/5851021045661696 Chrome might have implemented something in Chrome 68 which goes beyond the spec. But, some people in this thread are reporting that it's new as of Chrome 102, so, maybe it's not that at all, and is some other Chrome bug. Or maybe something changed in PayPal's code, around the same time Chrome/Edge 102 were released.

Discussing such issues in this thread is not helpful. If you are encountering an issue with your website in a given browser, where it worked in a previous version or works in other browser engines, you should post on the browser's issue tracker. For Chromium, that is https://crbug.com/.

Since we are generally shutting down this repository (see #72), I'll close this issue. Please feel free to have browser-specific discussions on browser-specific issue trackers.
